### PR TITLE
Add reusable `reviewing-openmc-code` skill and Copilot Review agent

### PR DESCRIPTION
**Pull request by paulromano**
# Description

This PR introduces AI-assisted code review tooling so that the review guidelines live in a single, tool-agnostic skill that can be used by Copilot, Claude Code, Codex, and similar agents. It makes the following updates:

- **`.claude/skills/openmc-code-review/SKILL.md`**: A reusable skill containing the code review procedure. Supports both *reviewer mode* (PR/branch review) and *self-review mode* (author preparing changes before submission). Note that I've chosen to put this in the `.claude` directory because it is recognized by both Claude Code and Copilot.
- **`.github/agents/Review.agent.md`**: A thin Github Copilot custom-agent wrapper that delegates to the `openmc-code-review` skill, providing a named "Review" persona in Copilot's agent dropdown.
- **`.github/copilot-instructions.md`**: Instructs Copilot to use the `openmc-code-review` skill when performing automated code reviews triggered directly on the GitHub repository.
- **`AGENTS.md`** (updated): Replaced the inline code-review guidance with a pointer to the `openmc-code-review` skill, avoiding duplication.

# Checklist

- [x] I have performed a self-review of my own code
- [x] <s>I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) (version 15) on any C++ source files (if applicable)</s>
- [x] <s>I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)</s>
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] <s>I have added tests that prove my fix is effective or that my feature works (if applicable)</s>

### CODE REVIEW

File: .claude/skills/openmc-code-review/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+**Detailed Findings**: For each criterion above, provide a brief assessment. Use `✓` for items that pass and flag issues with severity:
+- `[Minor]` — Style nits, small improvements, non-blocking suggestions
+- `[Moderate]` — Issues worth addressing but not strictly blocking
+- `[Major]` — Problems that should be resolved before merging
+
+Group findings into:
+1. **Blocking issues** — Would justify requesting changes before merge
+2. **Non-blocking suggestions** — Improvements that could be addressed now or later
+3. **Questions for the author** — Ambiguities or design choices worth clarifying
```
Comment (jtramm) [inline .claude/skills/openmc-code-review/SKILL.md]: I found that it would often ask questions that it was likely able to answer itself. E.g., "Can you confirm this is correct.." etc. The following proposal helps cut down on noise and forces it to think through things more.
```suggestion
3. **Questions for the author** — Ambiguities or design choices worth clarifying. Do not include questions that you are capable of answering yourself via deep analysis of the full repository. Only propose alternative designs if you have coded them up and can confirm they offer tangible improvements.
```

File: .claude/skills/openmc-code-review/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+## Determine Review Context
+
+1. **Identify what to review.** Determine the review context based on what the user provides or what can be inferred:
+   - **Default (no explicit context):** Compare the current branch against `develop` using `git diff develop...HEAD` to see only commits unique to the current branch.
+   - **User specifies a base branch or commit range:** Use that instead.
+   - **PR context provided by tooling:** Use the base/head refs supplied by the tool.
+2. **Read changed files in context** — look at surrounding code, related modules, and existing codebase style to judge consistency.
+3. **Determine review mode:**
+   - **Self-review mode** (author preparing changes before submission): Emphasize likely reviewer objections, missing tests/docs, API naming consistency, and quick fixes that would speed up review.
+   - **Reviewer mode** (default): Provide a standard structured review suitable for a code review or PR review.
```
Comment (jtramm) [inline .claude/skills/openmc-code-review/SKILL.md]: I'm wondering if having the two different modes is useful? It seems like the main reviewer mode should cover everything that the self-review mode would do as well (such that they're redundant), though perhaps you found some utility here? Have you noticed that the quality/quantity of the issues it flags changes significantly between the two modes? I have so far only tested it in review mode, but can do some more testing.

On the one hand it doesn't seem like it would hurt anything, on the other though, I have found that making the prompts more complex does potentially reduce the amount of time it will spend thinking on the actual review.

File: .claude/skills/openmc-code-review/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+## Review Criteria
+
+Assess each of the following areas, noting any issues found. If an area looks good, briefly confirm it passes.
+
+### Purpose and Scope
+- Do the changes have a clear, well-defined purpose?
+- Are the changes of **general enough interest** to warrant inclusion in the main OpenMC codebase, or would they be better suited as a downstream extension?
+
+### Correctness and Testing
+- Do the changes compile and appear functionally correct?
```
Comment (jtramm) [inline .claude/skills/openmc-code-review/SKILL.md]: Might strengthen language to get more effort out of the LLM potentially.

```suggestion
- Do the changes compile and can you confirm all logic to be functionally correct?
```

File: .claude/skills/reviewing-openmc-code/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+Apply repository-wide guidance from `AGENTS.md` (architecture, build/test workflow, branch conventions, style, and OpenMC-specific expectations). This skill adds a structured review process and rubric on top of that baseline knowledge.
+
+## Determine Review Context
+
+1. **Identify what to review.** Determine the review context based on what the user provides or what can be inferred:
+   - **Default (no explicit context):** Compare the current branch against `develop` using `git diff develop...HEAD` to see only commits unique to the current branch.
+   - **User specifies a base branch or commit range:** Use that instead.
+   - **PR context provided by tooling:** Use the base/head refs supplied by the tool.
+2. **Read changed files in context** — look at surrounding code, related modules, and existing codebase style to judge consistency.
```
Comment (jtramm) [inline .claude/skills/reviewing-openmc-code/SKILL.md:14]:
```suggestion
2. **Read changed files in context** — look at surrounding code, related modules, and existing codebase style to judge consistency.
3. **Explore repository** Given the context of the current PR, explore OpenMC to determine if there are any additional files you'll need to analyze given the multiple ways OpenMC can be run. You should read the (very short) main.cpp to get an idea of this as you begin. Read or Explore all files necessary to determine how this PR may intersect with different modes and run configurations: eigenvalue vs. fixed-source vs. volume vs. plotting modes; history-based vs. event-based transport; Monte Carlo vs. Random Ray; multigroup vs. continuous energy; particle restart and statepoint restart; statepoint/source I/O; surface source write; the C API accessors. Significant code is shared between the different run modes - you must understand how changes in the PR may affect other areas of the code not in the diff. Furthermore, ensure you have read enough to understand the global parallel status (both with MPI and OpenMP threading) of any regions of the diff. A minor change may look fine if assuming it runs in serial but a major bug if it is buried within a parallel region. 
```

Comment (jtramm) [reply .claude/skills/reviewing-openmc-code/SKILL.md:14]: The idea with this comment is to get it to spend a little more time up front exploring the repo so it pulls more of OpenMC into its context.

File: .claude/skills/reviewing-openmc-code/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+### Purpose and Scope
+- Do the changes have a clear, well-defined purpose?
+- Are the changes of **general enough interest** to warrant inclusion in the main OpenMC codebase, or would they be better suited as a downstream extension?
+
+### Correctness and Testing
+- Do the changes compile and appear functionally correct?
+- Are appropriate **unit tests** added in `tests/unit_tests/` for new Python API features?
+- Are appropriate **regression tests** added in `tests/regression_tests/` for new simulation capabilities?
+- Are edge cases and error conditions handled and tested?
```
Comment (jtramm) [inline .claude/skills/reviewing-openmc-code/SKILL.md:29]: Might also try some stronger language here to have it hopefully work hard to track down more bugs. 
```suggestion
- Are edge cases and error conditions handled and tested?
- Are all changes sound when considering that OpenMC runs in parallel with MPI and OpenMP (potentially at scale on massive exascale systems with billions or trillions of concurrent particles?)
- You have **UNLIMITED** time to complete the review -- the priority is to check every line and pull on every thread to ensure 100% correctness and that no errors are introduced to the repository. Do not rush to produce findings — a review that is thorough but slow is far more valuable than one that is fast but shallow. Finding one or two issues early does not mean the correctness analysis is done — keep looking with the same intensity through every category and every file. If you miss a single bug, the report quality is greatly diminished, so you need to pull on **EVERY** thread you find.
```

File: .claude/skills/openmc-code-review/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+- `[Major]` — Problems that should be resolved before merging
+
+Group findings into:
+1. **Blocking issues** — Would justify requesting changes before merge
+2. **Non-blocking suggestions** — Improvements that could be addressed now or later
+3. **Questions for the author** — Ambiguities or design choices worth clarifying
+
+**Recommended Action**:
+- *Reviewer mode*: One of — *Approve*, *Approve with minor suggestions*, *Request changes*, or *Reject* — with a brief justification.
+- *Self-review mode*: A prioritized list of pre-submission fixes — missing tests, undocumented parameters, naming inconsistencies, or anything else likely to slow down review.
```
Comment (jtramm) [inline .claude/skills/openmc-code-review/SKILL.md]: The report it generates is very official looking and well organized. Personally, I think I sort of unconsciously assume it must be authoritative and probably is capturing nearly everything, even though logically I know the limitations of LLMs and how complex the OpenMC codebase is. Maybe we'll tune this out after a few reviews, but it seems like it could be useful to give a warning here so as not to instill a false sense of confidence in the report. I've taken a crack at that, but you're welcome to axe this or go a different route.

```suggestion
- *Self-review mode*: A prioritized list of pre-submission fixes — missing tests, undocumented parameters, naming inconsistencies, or anything else likely to slow down review.

**Disclaimer**
Empirical evidence from Zeng et al. (https://arxiv.org/html/2509.01494v1) indicates LLMs (as of late 2025) can only find a very small percentage of bugs in PRs. OpenMC developer experience aligns well with the Zeng et al. findings. Thus, while this agent-based analysis is useful, it is likely only spotting a tiny fraction of the issues in complex PRs.
```

**Review by jtramm — APPROVED** [review]
This is great @paulromano!!! I had a lot of fun reviewing this and experimenting with making custom skills. I'm guessing this will be the first of many agentic skills for OpenMC in the near future.

I made a few suggestions for you to consider, but overall it seems like a great starting point for us.

For posterity, I've been benchmarking the claude code agentic review on my upcoming shared secondary branch draft PR (https://github.com/jtramm/openmc/tree/another_shared_secondary_branch). There's at least 10 legitimate logic bugs in there (not style/maintainability issues), but the default claude code review and this skill only find about 1 in 10 of the bugs on any given run. So, my own personal baseline expectation is that the agent will likely flag super obvious issues, but may only find about 10% of other more complex bugs. The default claude code `/review` skill doesn't do any better.

One trick I found is that due to random seeding, for complex PRs there is actually a huge divergence in the results you get (provided you run the review from clean/fresh contexts that aren't polluted from a previous review tool). As such, I've found that if I run multiple fully independent agentic reviews I'm able to expose way more issues than just running one. There's a paper here that shows this effect: https://arxiv.org/html/2509.01494v1 where you see significant improvements in PR reviewing if you collate 3-5 independent reviews, with greatly diminished returns going up to 10 or 20 reviews.

To that end, I may follow this up with a multi-agent skill that launches multiple such reviews in parallel so as to potentially uncover more bugs (as empirically this seems to be helpful, as discussed in https://arxiv.org/html/2509.01494v1, and in my findings as well). Sounds complex but claude already has an agent tool, so it's not too hard to get it to spawn a bunch of subagents to do parallel reviews (though it is more expensive!)

File: .claude/skills/openmc-code-review/SKILL.md (hunk 2 above)
Comment (paulromano) [reply .claude/skills/openmc-code-review/SKILL.md]: This is a good point, and along with your point about it looking "authoritative", I'm thinking it's probably best not to have the skill actually recommend approving/rejecting but rather just point out the issues. The human reviewer should be the one making the judgment call based on the information presented. In any case, I'm going to restructure to not have multiple modes since like you said, the work that it's doing should be the same.
